### PR TITLE
[Graph] Migrate Batch & Readout module to heterograph

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

* `dgl.batch` and `dgl.unbatch`
* Many functions under `dgl.readout`.

This PR will NOT pass CI. Need a final merge to pass everything.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/__init__.py [hunk 1]
```diff
@@ -20,9 +20,9 @@
 
 from .base import ALL, NTYPE, NID, ETYPE, EID
 from .readout import *
-from .batched_heterograph import *
+from .batch import *
 from .convert import *
-from .graph import DGLGraph, batch, unbatch
+from .graph import DGLGraph#, batch, unbatch
```
Comment (mufeili) [inline python/dgl/__init__.py:25]: Remove `#, batch, unbatch`?

File: python/dgl/backend/backend.py [hunk 2]
```diff
@@ -616,6 +616,11 @@ def split(input, sizes_or_sections, dim):
...
     ----------
     input : Tensor
+        Tensor to split.
+    sizes_or_sections : int, list[int]
+        Split sizes or sections.
+    dim : int
+        The dimension to split on.
 
     Returns
     -------
```
Comment (mufeili) [inline python/dgl/backend/backend.py:626]: Just for a note, need to do sanity check for `sizes_or_sections`.

File: python/dgl/batch.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+from . import utils
+
+__all__ = ['batch', 'unbatch']
+
+def batch(graphs, ndata=ALL, edata=ALL, *, node_attrs=None, edge_attrs=None):
+    r"""Batch a collection of ``DGLGraph``s into one graph for more efficient
+    graph computation.
+    
+    Each input graph becomes one disjoint component of the batched graph. The nodes
+    and edges are relabeled to be disjoint segments:
```
Comment (mufeili) [inline python/dgl/batch.py:18]: "relabeled to be disjoint segments" -> "relabeled as follows"

File: python/dgl/batch.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    performed on each graph individually, but become much more efficient
+    since they can be parallelized easily. This makes ``dgl.batch`` very useful
+    for tasks dealing with many graph samples such as graph classification tasks.
+
+    For heterograph inputs, they must share the same set of relations (i.e., node types
+    and edge types) and the function will perform batching on each relation one by one.
+    Thus, the result is also a heterograph and has the same set of relations as the inputs.
+    
+    The numbers of nodes and edges of the input graphs are accessible via the
+    :func:`DGLGraph.batch_num_nodes` and :func:`DGLGraph.batch_num_edges` attributes
```
Comment (mufeili) [inline python/dgl/batch.py:37]: I presume `DGLGraph.batch_num_nodes` and `DGLGraph.batch_num_edges` are now functions with node type/edge type introduced?

File: python/dgl/batch.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            
+    Batch heterographs
+    
+    >>> hg1 = dgl.heterograph({('user', 'plays', 'game') : (th.tensor([0, 1]), th.tensor([0, 0]))})
+    >>> hg2 = dgl.heterograph({('user', 'plays', 'game') : (th.tensor([0, 0, 0]), th.tensor([1, 0, 2]))})
+    >>> bhg = dgl.batch([hg1, hg2])
+    >>> bhg
+    Graph(num_nodes={'user': 3, 'game': 4},
+          num_edges={('user', 'plays', 'game'): 5},
+          metagraph=[('drug', 'game')])
```
Comment (mufeili) [inline python/dgl/batch.py:135]: `'drug'` -> `'user'`

File: python/dgl/utils.py [hunk 6]
```diff
@@ -739,3 +739,48 @@ def check_all_same_device(glist, name):
...
+        Name of this feature dict.
+    """
+    if len(feat_dict_list) == 0:
+        return
+    for fdict in feat_dict_list:
+        for k in keys:
+            t1 = feat_dict_list[0][k]
+            t2 = fdict[k]
+            if F.dtype(t1) != F.dtype(t2) or F.shape(t1)[1:] != F.shape(t2)[1:]:
+                raise DGLError('Expect all feature "{}" to have the data type'
```
Comment (mufeili) [inline python/dgl/utils.py:784]: "the data type" -> "the same data type"?

File: python/dgl/heterograph.py [hunk 7]
```diff
@@ -631,6 +633,52 @@ def get_etype_id(self, etype):
         if etid is None:
             raise DGLError('Edge type "{}" does not exist.'.format(etype))
         return etid
+    #################################################################
+    # Batching
+    #################################################################
+    @property
+    def batch_size(self):
```
Comment (mufeili) [inline python/dgl/heterograph.py:640]: Doc?

File: python/dgl/heterograph.py [hunk 8]
```diff
@@ -631,6 +633,52 @@ def get_etype_id(self, etype):
...
+        return len(self.batch_num_nodes(self.ntypes[0]))
+
+    def batch_num_nodes(self, ntype=None):
```
Comment (mufeili) [inline python/dgl/heterograph.py:643]: The behavior seems to be different from the one described in `batch.py`, where `batch_num_nodes` returns a dictionary in the case of multiple node types.

Comment (mufeili) [inline python/dgl/heterograph.py:643]: Add doc

File: python/dgl/heterograph.py [hunk 9]
```diff
@@ -631,6 +633,52 @@ def get_etype_id(self, etype):
...
+        return self._batch_num_nodes[ntype]
+
+    def set_batch_num_nodes(self, val):
+        if not isinstance(val, Mapping):
+            if len(self.ntypes) != 1:
+                raise DGLError('Must provide a dictionary when there are multiple node types.')
+            val = {self.ntypes[0] : val}
+        self._batch_num_nodes = val
+
+    def batch_num_edges(self, etype=None):
```
Comment (mufeili) [inline python/dgl/heterograph.py:663]: Same as `batch_num_nodes`

**Comment by mufeili** [comment]
Overall LGTM, some minor comments

File: python/dgl/readout.py [hunk 10]
```diff
@@ -5,79 +5,44 @@
...
-
-def sum_nodes(graph, feat, weight=None):
-    """Sums all the values of node field :attr:`feat` in :attr:`graph`, optionally
-    multiplies the field by a scalar node field :attr:`weight`.
+    The function is commonly used as a *readout* function on a batch of graphs
+    to generate graph-level representation. Thus, the result tensor shape
+    depends on the batch size of the input graph. Given a graph of batch size
+    :math:`B`, and a feature size of :math:`D`, the result shape will be
+    :math:`(B, D)`, with each row being the aggregated node features of each
+    graph.
```
Comment (mufeili) [inline python/dgl/readout.py:25]: Briefly describe the behavior in the case of multiple node types and give an example?

File: python/dgl/readout.py [hunk 11]
```diff
@@ -88,65 +53,73 @@ def sum_nodes(graph, feat, weight=None):
...
+def readout_edges(graph, feat, weight=None, *, op='sum', etype=None):
+    """Sum the edge feature :attr:`feat` in :attr:`graph`, optionally
+    multiplies it by a edge :attr:`weight`.
+
+    The function is commonly used as a *readout* function on a batch of graphs
+    to generate graph-level representation. Thus, the result tensor shape
+    depends on the batch size of the input graph. Given a graph of batch size
+    :math:`B`, and a feature size of :math:`D`, the result shape will be
+    :math:`(B, D)`, with each row being the aggregated edge features of each
+    graph.
```
Comment (mufeili) [inline python/dgl/readout.py:101]: Briefly describe the behavior in the case of multiple edge types and give an example?

**Comment by mufeili** [comment]
Overall LGTM, some minor comments

**Comment by jermainewang** [comment]
@mufeili For docstring suggestions, let's directly update them in the hackmd and post them later.

**Comment by mufeili** [comment]
> @mufeili For docstring suggestions, let's directly update them in the hackmd and post them later.

Ok

**Comment by classicsong** [comment]
Some comments
Others LGTM.

File: python/dgl/batch.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+    {'user' : tensor([2, 1]), 'game' : tensor([1, 3])}
+    >>> bhg.batch_num_edges()
+    {('user', 'plays', 'game') : tensor([2, 3])}
+    
+    See Also
+    --------
+    unbatch
+    """
+    if len(graphs) == 0:
+        raise DGLError('The input list of graphs cannot be empty.')
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/batch.py:148]: I think the correct behavior would be `return graphs[0].clone()` in this case.

**Pull request merged**